distogram/__init__.py

```diff
@@ -1,240 +1,203 @@
+__author__ = """Romain Picard"""
+__email__ = 'romain.picard@oakbits.com'
+__version__ = '1.7.0'
+
 import math
-from functools import reduce
 from bisect import bisect_left
+from functools import reduce
+from operator import itemgetter
+from typing import List
+from typing import Optional
+from typing import Tuple
 
-__author__ = """Romain Picard"""
-__email__ = 'romain.picard@oakbits.com'
-__version__ = '1.6.0'
+EPSILON = 1e-5
+Bin = Tuple[float, int]
 
 
 # bins is a tuple of (cut point, count)
 class Distogram(object):
-    '''Compressed representation of a distribution
-    '''
+    """ Compressed representation of a distribution.
+    """
     __slots__ = 'bin_count', 'bins', 'min', 'max', 'diffs', 'min_diff', 'weighted_diff'
 
-    def __init__(self, bin_count=100, weighted_diff=False):
-        '''Creates a new Distogram object
+    def __init__(self, bin_count: int = 100, weighted_diff: bool = False):
+        """ Creates a new Distogram object
 
         Args:
             bin_count: [Optional] the number of bins to use.
+            weighted_diff: [Optional] Whether to use weighted bin sizes.
 
         Returns:
             A Distogram object.
-        '''
-        self.bin_count = bin_count
-        self.bins = []
-        self.min = None
-        self.max = None
-        self.diffs = None
-        self.min_diff = None
-        self.weighted_diff = weighted_diff
+        """
+        self.bin_count: int = bin_count
+        self.bins: List[Bin] = list()
+        self.min: Optional[float] = None
+        self.max: Optional[float] = None
+        self.diffs: Optional[List[float]] = None
+        self.min_diff: Optional[float] = None
+        self.weighted_diff: bool = weighted_diff
 
 
-def _linspace(start, stop, num):
+def _linspace(start: float, stop: float, num: int) -> List[float]:
     if num == 1:
-        return stop
-    h = (stop - start) / float(num)
-    values = [start + h * i for i in range(num+1)]
+        return [stop]
+    step = (stop - start) / float(num)
+    values = [start + step * i for i in range(num + 1)]
     return values
 
 
-def _moment(x, counts, c, n):
-    m = []
-    for i in range(len(x)):
-        m.append(counts[i]*(x[i]-c)**n)
+def _moment(x: List[float], counts: List[float], c: float, n: int) -> float:
+    m = (ci * (v - c) ** n for i, (ci, v) in enumerate(zip(counts, x)))
```

Comment:
using an iterator lit this is much faster than a for-loop because the interpreter runs it as a single line.